File: t9/proto/cache.py
from keyTrie import *

class CacheNode(TrieNode):
	def __init__(self, parentNode):
		TrieNode.__init__(self)
		self.parentNode = parentNode
	def setParent(self, parentNode):
		self = CacheNode(parentNode)
		
class CacheTrie(Trie):
	def __init__(self):
		Trie.__init__(self, "none")
		self.root = CacheNode(None)
	
	def insert(self, word, frequency):
		parentNode = None
		node = self.root
		key = 0
		for c in word:
			key = keys[c]
			if not node.hasChildren(key):
				node.putNode(key, CacheNode(parentNode))
			parentNode = node
			node = node.getBranch(key)
		node.setAsLeaf()
		node.words.append([word, frequency])
		node.words = sorted(node.words, key = itemgetter(1))
		
	def getSuggestions(self, keySequence, suggestionDepth):
		return self.__getSuggestions__(keySequence, suggestionDepth)
		
class Cache:
	def __init__(self, sizeLimit):
		self.sizeLimit = sizeLimit
		self.cacheList = []
		self.cacheTrie = CacheTrie()

	def getSuggestions(self, keySequence, suggestionDepth):
		return self.cacheTrie.getSuggesions(keySequence, suggestionDepth)

	def update(self, chosenWord, frequency, keySequence):
		oldIndex = -1
		for i, word in enumerate(self.cacheList):
			if word == chosenWord:
				oldIndex = i
				break
		# if chosenWord is in the cache, move it to front,
		# and update frequency in cacheTrie
		if oldIndex != -1:
			self.cacheList.insert(0, self.cacheList.pop(oldIndex))
			self.updateFrequency(chosenWord, keySequence)
		else:
			self.insert(chosenWord, frequency)
			
	def updateFrequency(self, chosenWord, keySequence):
		prefix = self.getPrefixLeaf(keySeq)
		if self.wordExists(chosenWord, keySeq):
			for i in prefix[0].words:
				if i[0] == chosenWord:
					i[1] += 1
					break

	# Only call from update() so that we've already checked
	# to see if chosenWord is in the cache
	def insert(self, word, frequency):
		# if @ capacity
		if len(self.cacheList) == self.sizeLimit:
			self.pruneOldest()
		# put most recent word at beginning
		if self.cacheList:
			self.cacheList[0] = word
		self.cacheTrie.insert(word, frequency)

	def pruneOldest(self):
		self.pruneWord(cacheList[len(cacheList) - 1])

	def pruneWord(self, wordToPrune):
		nodeToPrude = CacheTrie.getPrefixNode(wordToPrune)
		# If wordToPrune is a prefix with other children, just remove this one word from the word list of nodeToPrune
		if nodeToPrune.children:
			nodeToPrune.words.remove(wordToPrune)
			return
		else:
			self.pruneNode(nodeToPrune)

	def pruneNode(self, nodeToPrune):
		# parent has no keys in children other than the target's key
		if len(nodeToPrune.parentNode.children) == 1:
			self.pruneNode(nodeToPrune.parentNode)
		else:
			nodeToPrune = None

	def clear(self):
		self = Cache(self.sizeLimit)

File: calc_pad/CalculatorKeyboard-master/Calculator/cache.swift
internal class CacheNode: TrieNode {
	let parentNode: CacheNode?
	init(parentNode: CacheNode?) {
		self.parentNode = parentNode
	}
    override func getBranch(key: Int) -> CacheNode {
        return self.children[key]! as! CacheNode
    }
}

internal class CacheTrie {
    
    let root: CacheNode
    
    internal class DeeperSuggestion {
        var deeperSuggestions: Array<Array<WordWeight>> = [[WordWeight]]()
    }
    
    init() {
        self.root = CacheNode(parentNode: nil)
    }
    
	internal func insert(word: String, weight: Int) {
		var parentNode = CacheNode(parentNode: nil)
		var node = self.root
		var key = 0
		for c in word.characters {
			key = lettersToDigits[String(c)]!
            if !node.hasChild(key: key) {
				node.putNode(key: key, nodeToInsert: CacheNode(parentNode: parentNode))
			}
			parentNode = node
			node = node.getBranch(key: key)
		}
		node.setAsLeaf()
        node.wordWeights.append(WordWeight(word: word, weight: weight))
		node.wordWeights = node.wordWeights.sorted(by: {$0.weight > $1.weight})
	}
    
    internal func getSuggestions(keySequence : [Int], suggestionDepth : Int) -> [String] {
        var suggestions = [String]()
        let prefixNode: TrieNode? = self.getPrefixNode(keySequence: keySequence)
        
        if prefixNode != nil {
            for wordWeight in prefixNode!.wordWeights {
                suggestions.append(wordWeight.word)
            }
            
            if suggestionDepth > 1 {
                var deeperSuggestions = DeeperSuggestion()
                // deeperSuggestions is a classs, so it is passed by reference
                // After the call to getDeeperSuggestions, deeperSuggestions
                // will be a list of lists of words, each list being full of
                // words of one character longer in length
                self.getDeeperSuggestions(root: prefixNode!,
                                          maxDepth:
                    keySequence.count + suggestionDepth,
                                          deeperSuggestions: deeperSuggestions)
                
                for level in deeperSuggestions.deeperSuggestions {
                    for wordWeight in level {
                        suggestions.append(wordWeight.word)
                    }
                }
            }
        }
        
        return suggestions
    }
    
    internal func getDeeperSuggestions(root : TrieNode, maxDepth : Int,
                                       deeperSuggestions: DeeperSuggestion) {
        self.traverse(root: root, depth: 0, maxDepth: maxDepth, deepSuggestions: deeperSuggestions)
        for (level, suggestions) in deeperSuggestions.deeperSuggestions.enumerated() {
            if suggestions.count > 0 {
                deeperSuggestions.deeperSuggestions[level] =
                    suggestions.sorted(by: {$0.weight > $1.weight})
            }
        }
    }
    
    internal func traverse(root : TrieNode, depth : Int, maxDepth : Int,
                           deepSuggestions : DeeperSuggestion) {
        if (depth < maxDepth && depth > 0) {
            for wordWeight in root.wordWeights {
                deepSuggestions.deeperSuggestions[depth-1].append(wordWeight)
            }
        }
        
        if depth == maxDepth || root.children.count == 0 {
            return
        }
        
        for (key, _) in root.children {
            self.traverse(root: root.children[key]!, depth: depth+1,
                          maxDepth: maxDepth, deepSuggestions: deepSuggestions)
        }
    }
    
    internal func getPrefixLeaf(keySequence : [Int]) -> (CacheNode?, Bool) {
        var node: CacheNode? = self.root
        var prefixExists = true
        
        for (i, key) in keySequence.enumerated() {
            if node!.hasChild(key: key) {
                node = node!.getBranch(key: key)
            }
            else {
                if i == keySequence.count - 1 {
                    prefixExists = true
                }
                else {
                    prefixExists = false
                    node = nil
                    return (node, prefixExists)
                }
            }
        }
        return (node, prefixExists)
    }
    
    internal func getPrefixNode(keySequence : [Int]) -> CacheNode? {
        let (node, prefixExists) = self.getPrefixLeaf(keySequence: keySequence)
        if prefixExists {
            return node
        }
        else {
            return nil
        }
    }
    
    internal func wordExists(word : String, keySequence: [Int]) -> Bool {
        let (node, _) = self.getPrefixLeaf(keySequence: keySequence)
        if node != nil {
            if node!.isLeaf() {
                for wordWeight in node!.wordWeights {
                    if wordWeight.word == word {
                        return true
                    }
                }
                return false
            }
            else {
                return false
            }
        }
        else {
            return false
        }
    }
}

public class Cache {
	let sizeLimit: Int
	var cacheList: [String]
	var cacheTrie = CacheTrie()
	init(sizeLimit: Int) {
		self.sizeLimit = sizeLimit
		self.cacheList = []
	}
    
	func getSuggestions(keySequence: [Int], suggestionDepth: Int) -> [String] {
		return self.cacheTrie.getSuggestions(keySequence: keySequence,
		                                    suggestionDepth: suggestionDepth)
	}
    
	func update(chosenWord: String, weight: Int, keySequence: [Int]) {
		var oldIndex = -1
		for (i, word) in cacheList.enumerated() {
			if word == chosenWord {
				oldIndex = i
				break
			}
		}
		// if chosenWord is in the cache, move it to front,
		// and update weight in cacheTrie
		if oldIndex != -1 {
            let lastWord = self.cacheList[oldIndex]
            self.cacheList.remove(at: oldIndex)
            self.cacheList.insert(lastWord, at: 0)
			self.updateWeight(word: chosenWord)
		}
		else {
			self.insert(word: chosenWord, weight: weight)
		}
	}
    
	internal func updateWeight(word: String) {
        let keySequence = getKeySequence(word: word)
        let prefixNode = cacheTrie.getPrefixLeaf(keySequence: keySequence).0
		if cacheTrie.wordExists(word: word, keySequence: keySequence) {
			for wordWeight in prefixNode!.wordWeights {
				if wordWeight.word == word {
                    wordWeight.weight += 1
					break
				}
			}
		}
	}
    
	// Only call from update() so that we've already checked
 	// to see if chosenWord is in the cache
	internal func insert(word: String, weight: Int) {
		// if @ capacity
		if self.cacheList.count == self.sizeLimit {
			self.pruneOldest()
		}
		// put most recent word at beginning
		if self.cacheList.count > 0 {
			self.cacheList[0] = word
		}
		self.cacheTrie.insert(word: word, weight: weight)
	}
    
	internal func pruneOldest() {
		self.pruneWord(wordToPrune: cacheList[cacheList.count - 1])
	}
    
	internal func pruneWord(wordToPrune: String) {
        let keySequnce = getKeySequence(word: wordToPrune)
        var nodeToPrune = self.cacheTrie.getPrefixNode(keySequence: keySequnce)
		// If wordToPrune is a prefix with other children, just remove this one word from the word list of nodeToPrune
		if (nodeToPrune?.children.count)! > 0 {
            var wordIndex = 0
            for (i, wordWeight) in (nodeToPrune?.wordWeights.enumerated())! {
                if wordWeight.word == wordToPrune {
                    wordIndex = i
                }
            }
			nodeToPrune?.wordWeights.remove(at: wordIndex)
			return
		}
		else {
            while nodeToPrune?.parentNode?.children.count == 1 {
                nodeToPrune = nodeToPrune?.parentNode
            }
            nodeToPrune = nil
		}
	}
}
